feat: ✨ add storybook to cli

lloydrichards opened the pull request:
## Goals/Scope

This is extension of PR #11 regarding adding stories for storybook through the CLI. A lot of work was already done by [XavierGeerinck](https://github.com/XavierGeerinck) to get this going, but the PR seems to have been lost despite renewed interest and push to get this feature added. I've created this renewed PR with the hope of getting some attention and momentum in adding this to the CLI.

For the sake of narrowing the scope of this PR, i've removed the typography component so we can focus on just storybook.

## Description

There are two main areas that have been changed in this PR, first is the addition of a storybook folder in the /registry which contain `*.story.tsx` files for most of the component (some are still missing). Second is I've added an additional option to the `component.tsx` which should allow for the stories to be loaded in through the CLI when adding a component (still need to test)

<img width="1423" alt="Screenshot 2024-02-13 at 20 18 41" src="https://github.com/shadcn-ui/ui/assets/59728961/85480dee-cb65-4c5e-a27d-def69ea3a89f">

### Component Stories

The stories are added to the `registry` folder in the `*.story.tsx` files. While most of the components follow the examples in the shadcn/ui docs, some of the examples needed to be simplified or broken up into multiple stories to make them more readable. Below you will find a list of the components and their stories.
<details>
  <summary>Expand Component Table</summary>

| Checkbox | Name            |                            Link to [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/) |
| :------: | --------------- | ---------------------------------------------------------------------------------------------------------: |
|    ✅    | accordion       |      [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-accordion--docs) |
|    ✅    | alert           |          [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-alert--docs) |
|    ✅    | alert-dialog    |    [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-alertdialog--docs) |
|    ✅    | aspect-ratio    |    [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-aspectratio--docs) |
|    ✅    | avatar          |         [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-avatar--docs) |
|    ✅    | badge           |          [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-badge--docs) |
|    ✅    | breadcrumb      |     [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-breadcrumb--docs) |
|    ✅    | button          |         [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-button--docs) |
|    ✅    | calendar        |       [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-calendar--docs) |
|    ✅    | card            |           [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-card--docs) |
|    ✅    | carousel        |       [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-carousel--docs) |
|    ✅    | checkbox        |       [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-checkbox--docs) |
|    ✅    | collapsible     |    [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-collapsible--docs) |
|    ✅    | command         |        [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-command--docs) |
|    ✅    | context-menu    |    [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-contextmenu--docs) |
|    ✅    | dialog          |         [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-dialog--docs) |
|    ✅    | drawer          |         [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-drawer--docs) |
|    ✅    | dropdown-menu   |   [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-dropdownmenu--docs) |
|    ✅    | form            |           [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-form--docs) |
|    ✅    | hover-card      |      [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-hovercard--docs) |
|    ✅    | input-otp       |       [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-inputotp--docs) |
|    ✅    | input           |          [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-input--docs) |
|    ✅    | label           |          [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-label--docs) |
|    ✅    | menubar         |        [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-menubar--docs) |
|    ✅    | navigation-menu | [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-navigationmenu--docs) |
|    ✅    | pagination      |     [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-pagination--docs) |
|    ✅    | popover         |        [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-popover--docs) |
|    ✅    | progress        |       [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-progress--docs) |
|    ✅    | radio-group     |     [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-radiogroup--docs) |
|    ✅    | resizable       |      [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-resizable--docs) |
|    ✅    | scroll-area     |     [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-scrollarea--docs) |
|    ✅    | select          |         [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-select--docs) |
|    ✅    | separator       |      [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-separator--docs) |
|    ✅    | sheet           |          [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-sheet--docs) |
|    ✅    | skeleton        |       [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-skeleton--docs) |
|    ✅    | slider          |         [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-slider--docs) |
|    ✅    | sonner          |         [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-sonner--docs) |
|    ✅    | switch          |         [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-switch--docs) |
|    ✅    | table           |          [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-table--docs) |
|    ✅    | tabs            |           [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-tabs--docs) |
|    ✅    | textarea        |       [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-textarea--docs) |
|    ✅    | toast           |          [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-toast--docs) |
|    ✅    | toggle          |         [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-toggle--docs) |
|    ✅    | toggle-group    |    [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-togglegroup--docs) |
|    ✅    | tooltip         |        [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-tooltip--docs) |

| Bonus | Name                     |                                                                                                    Link to Stories |
| :---: | ------------------------ | -----------------------------------------------------------------------------------------------------------------: |
|  ⭐️  | color - functional       |       [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/story/design-color--functional) |
|  ⭐️  | color - tailwind         |         [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/story/design-color--tailwind) |
|  ⭐️  | radius                   |            [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/story/design-radius--core) |
|  ⭐️  | shadow                   |            [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/story/design-shadow--core) |
|  ⭐️  | padding                  |           [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/story/design-padding--core) |
|  ⭐️  | typography - font family | [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/story/design-typography--font-family) |
|  ⭐️  | typography - font size   |   [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/story/design-typography--font-size) |
|  ⭐️  | typography - font weight | [Storybook](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/story/design-typography--font-weight) |
</details>

## Comments

I'm very happy for support with getting this PR through so if anyone would like to review the stories, or add new ones, or help with upgrading the CLI, please join the conversation in the comments 🙇

_if you would like to add a missing story, you can create the `.story.tsx` file in the `/apps/www/registry/stories/` directory and add the file to the `apps/www/public/registry/index.json` for the correct component._

---

- [x] cc: @XavierGeerinck @Integrayshaun

csartor commented:
I believe!!!

Jared-Dahlke commented:
@lloydrichards did you have any trouble with the code blocks showing corrupted code? Happens to me with <Source/> and just the auto docs
![image](https://github.com/shadcn-ui/ui/assets/32936978/781ce4da-0aff-4103-8cde-97a8149a63ac)

lloydrichards commented:
> @lloydrichards did you have any trouble with the code blocks showing corrupted code? Happens to me with `<Source/>` and just the auto docs ![image](https://user-images.githubusercontent.com/32936978/271041011-781ce4da-0aff-4103-8cde-97a8149a63ac.png)

I had a look and things loook correct in the code block for my dev env, but i've gone and upgraded storybook to v7.4.5 incase there was some bug in there somewhere.  are you using `pnpm storybook` to spin it up?

Jared-Dahlke commented:
> > @lloydrichards did you have any trouble with the code blocks showing corrupted code? Happens to me with `<Source/>` and just the auto docs
> 
> I had a look and things loook correct in the code block for my dev env, but i've gone and upgraded storybook to v7.4.5 incase there was some bug in there somewhere. are you using `pnpm storybook` to spin it up?

i implemented a lot of stories myself before i found your repo. Now i cloned yours and am running it and got this error:

![image](https://github.com/shadcn-ui/ui/assets/32936978/2081c87f-6a2f-4468-8f8e-45e224570ab2)

just fyi. Seems like awesome work though. I've also built tests for about half of the components so far. i will have to write tests for all of them soon

lloydrichards commented:
> i implemented a lot of stories myself before i found your repo. Now i cloned yours and am running it and got this error:

I think the issue is using npm vs pnpm.  i usually use yarn and was running into issues with yarn install, but noticed the pnpm.lock so tried using pnpm instead.

Jared-Dahlke commented:
@lloydrichards what do you think is the easiest way to add the import code block on each story at the top. like this:

```jsx
import { Button } from 'mylibrary/button'`
```

Right now i'm having to make a .mdx for each story and put that at the top manually.

Integrayshaun commented:
@lloydrichards This looks great!!! 😁 is there anything that you need from me or the Storybook team to wrap up?

lloydrichards commented:
> is there anything that you need from me or the Storybook team to wrap up?

One piece of the puzzle that is still missing is bring able to infer the arg types better for the components.  Normally in a story the Meta<typeof Component> does a pretty good job building out the controls, but i've noticed with the Radix components that this only works if you specify a prop in the Story.args.  Do you know if there is a better way to infer the argTypes, otherwise its a manual process of picking which props are most useful for each story

lloydrichards commented:
@shadcn Would it be possible to have a look and see if the changes im proposing here to the CLI and the inclusion o Storybook is inline with you roadmap with shadch/ui?  So far I've achieved the core of what i want to do with this PR:
- [x] write storybook files for all ui components
- [x] integrate the stories into the CLI so they can be loaded when added
- [x] keep the PR up to date with the latest main and storybook versions

next steps for me really depend if there is any pushback from your side or any requests for changes.  just let me know or if you don't have time, maybe ping another maintainer so I  can try test the CLI? 🙏

kylegach commented:
> One piece of the puzzle that is still missing is bring able to infer the arg types better for the components. Normally in a story the Meta does a pretty good job building out the controls, but i've noticed with the Radix components that this only works if you specify a prop in the Story.args. Do you know if there is a better way to infer the argTypes, otherwise its a manual process of picking which props are most useful for each story

@lloydrichards —

This is most likely because [Storybook, by default, filters out all props sourced from `node_modules`](https://storybook.js.org/docs/react/api/main-config-typescript#reactdocgentypescriptoptions) (otherwise, you end up with every prop from, say, `@types/react`, which includes all valid HTML attributes).

Try adjusting the prop filter like so:

```ts
// .storybook/main.ts
import type { StorybookConfig } from "@storybook/nextjs"

const config: StorybookConfig = {
  // ...
  typescript: {
    reactDocgen: 'react-docgen-typescript',
    reactDocgenTypescriptOptions: {
      shouldExtractLiteralValuesFromEnum: true,
      // 👇 Adjusted prop filter
      propFilter: (prop) =>
        prop.parent ? !/node_modules\/(?!@radix-ui)/.test(prop.parent.fileName) : true,
    },
  },
}
export default config
```

shadcn commented:
@lloydrichards This is incredible work. 

I'd like to take a closer look at this (mostly to understand what maintenance effort this will require). 

I'll add this to my list and get back to it early next week.

Integrayshaun commented:
@shadcn If you have any questions for the Storybook team or ever need a hand, we're always around

kylegach commented:
@Jared-Dahlke — Wanted to chime in to make sure you're aware of [Storybook's interaction tests](https://storybook.js.org/docs/react/writing-tests/interaction-testing#write-an-interaction-test)?

That test you shared could be done entirely within a play function, which is part of the story. It also runs in a real browser instead of JSDom.

Then you could [debug each step interactively](https://storybook.js.org/docs/react/writing-tests/interaction-testing#interactive-debugger).

Automating is done via [test-runner](https://storybook.js.org/docs/react/writing-tests/test-runner), which will smoke test all stories and run any play functions that may exist.

The end result is similar (except for the browser vs. JSDom environment and better debugging), but with far less boilerplate.

lloydrichards commented:
Just want to throw my thoughts out for including testing with the stories:

Personally I do use storybook's interaction testing on a lot of my components, but I think it adds an extra barrier to entry when imagining these stories as being part of shadcn/ui CLI.  I love that shadcn/ui takes away a lot of the boilplate needed to work with radix primatives and doesn't bloat the files with things I might never need, while still allowing me to expand on the functionality if need.

In this way I would like to try keep the stories as simple as possible with no extra configuration to the `.storybook/main.ts` or `.storybook/preview.tsx` files.  Anything past the storybook essentials should be given to the user to implement.

linkb15 commented:
@lloydrichards If we want to add a new style, would the storybook styling and examples need to be updated as well? 

since each styles are with different icons and sizes.

lloydrichards commented:
> would the storybook styling and examples need to be updated as well?

Technically there shouldn't be any styling in the stories since they pull the components in from the respective file.  With either default or New York or their own overrides, the same story will work.  It could be a good thing to review or keep in mind though, not to create too many custom elements in these stories to keep them generic of the theming

stramel commented on apps/www/.storybook/main.ts:
```diff
@@ -0,0 +1,24 @@
+    "@storybook/addon-links",
+    "@storybook/addon-essentials",
+    "@storybook/addon-interactions",
+    "@storybook/addon-styling",
```
`@storybook/addon-styling` is deprecated in favor of `@storybook/addon-styling-webpack`

https://storybook.js.org/addons/@storybook/addon-styling
https://github.com/storybookjs/addon-styling-webpack

Integrayshaun replied:
Good catch, @stramel. In this case, because the `@storybook/nextjs` framework package is being used, we won't need either package to get tailwind working 😁

lloydrichards replied:
thanks for the catch! I was upgrading to storybook v7.5.2 anyway so I took the opportunity to remove these 👍

lloydrichards commented on apps/www/registry/stories/tooltip.stories.tsx:
```diff
@@ -0,0 +1,36 @@
+
+export const Base: Story = {
+  render: (args) => (
+    <Tooltip {...args}>
```
Thanks for catching this! just patched this up

lloydrichards commented on apps/www/registry/stories/select.stories.tsx:
```diff
@@ -0,0 +1,62 @@
+export const Base: Story = {
+  render: (args) => (
+    <Select {...args}>
+      <SelectTrigger className="w-[180px]">
```
From the Radix documentation on for Select ([docs](https://www.radix-ui.com/primitives/docs/components/select)) it seems like this shouldn't be needed: 
>Adheres to the [ListBox WAI-ARIA design pattern](https://www.w3.org/WAI/ARIA/apg/patterns/listbox).  
><sub>[link to text](https://www.radix-ui.com/primitives/docs/components/select#:~:text=Adheres%20to%20the%20ListBox%20WAI%2DARIA%20design%20pattern.)</sub>

Though its possible to add a `<Label />` to the Select, I think for this simple of an example its more to showcase the interaction with the select+dropdown.  This could maybe be brought up with the documentation and example?

lloydrichards commented:
Have made the stories for the new 0.5.0 components based off the documentations.  Think there might be some interesting additional stories that could be added to show the examples, but left out for the time being.  Anyone want to review or add suggestions, i'm all ears 👂 .

pmm26 commented:
Looks like it's going to take a long time before we see it deployed :/

shadcn commented:
@lloydrichards @pmm26 I've started testing it. It's a pretty big PR and I have to figure out to maintain it in the long term...etc. Definitely want to land this though... :+1:

Integrayshaun commented:
@shadcn I'm happy to help own the maintenance of the Storybook stuff ❤️

lloydrichards commented:
> It's a pretty big PR and I have to figure out to maintain it in the long term...etc.

Let me know if you need anything from me to help with this process.  I was thinking about building a page in the docs that can quickly overview the storybook feature, but I understand that will only increase the amount to review so can draft it up as a new PR if that makes things easier.  Excited to see this land, so willing to help and commit to any maintenance in the future like @Integrayshaun.  🙏  thanks!

Integrayshaun commented:
@lloydrichards I'll take a stab at reviewing the work this afternoon 👍

lloydrichards commented:
Did some refactoring to add some documentation in the form of JSDocs as well reduced repetition through Story arg composition.  Am still thinking about expanding the argType to be more descriptive of the properties, especially for components where the ArgTable generation isn't automatic or for others that are not very descriptive.
<img width="541" alt="Screenshot 2024-01-22 at 23 18 14" src="https://github.com/shadcn-ui/ui/assets/59728961/9d41cfd2-3c29-4355-ba84-188d0704f1e1">

> @lloydrichards I'll take a stab at reviewing the work this afternoon 👍

If you see any other places where the story composition could be improved, i'm open for improving them 🙏 

---
Something I noticed while adding some more explicit storied from the RadixUI props was that there doesnt seem to be default styling for disabled components like `Accordion`, `RadioGroup`, and `Checkbox` like there is `Select` or `Switch`.  Maybe someone feels like making a new PR for this fix?
<img width="552" alt="Screenshot 2024-01-22 at 23 20 36" src="https://github.com/shadcn-ui/ui/assets/59728961/56a79eec-fc6d-4a55-b808-5f01a5dffed3">

kylegach commented:
Hey, folks. Love this effort! @ShaunEvening has asked me to take over as the "Storybook contributor voice" and I'm happy to do so. Will review this in the next few days!

kylegach commented on apps/www/.storybook/main.ts:
```diff
@@ -0,0 +1,22 @@
+  stories: [
+    "../registry/stories/**/*.mdx",
+    "../registry/stories/**/*.stories.@(js|jsx|ts|tsx)",
+    "./theme/**/*.stories.@(js|jsx|ts|tsx|mdx)",
```
[Storybook 8 will remove support for `*.stories.mdx`.](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#dropping-support-for-storiesmdx-csf-in-mdx-format-and-mdx1-support)

I only see one file using that filename convention, `./theme/colors.stories.mdx`, and it doesn't contain any stories, so you can make the following changes to maintain the current structure and be ready for SB 8:

1. Update this line like so:

   ```diff
   - "./theme/**/*.stories.@(js|jsx|ts|tsx|mdx)",
   + "./theme/**/*.mdx",
   + "./theme/**/*.stories.@(js|jsx|ts|tsx)",
   ```

   (Technically, that second line isn't necessary for the current content, but if anyone ever adds a `*.stories.ts` file in the `theme` directory, this will ensure it works.)

1. Rename `./theme/colors.stories.mdx` to `colors.mdx`

kylegach commented on apps/www/registry/stories/aspect-ratio.stories.tsx:
```diff
@@ -0,0 +1,69 @@
+ * Use the `1:1` aspect ratio to display a square image.
+ */
+export const Square: Story = {
+  render: Default.render,
```
If you're going to re-use the same `render` across every story for a component, you can define it on the `meta` instead.

kylegach commented on apps/www/registry/stories/accordion.stories.tsx:
```diff
@@ -0,0 +1,66 @@
+ * A vertically stacked set of interactive headings that each reveal a section
+ * of content.
+ */
+const meta: Meta<typeof Accordion> = {
+  title: "ui/Accordion",
+  component: Accordion,
+  tags: ["autodocs"],
+  argTypes: {},
+}
+
+export default meta
+
+type Story = StoryObj<typeof Accordion>
```
Given that you're using TS 4.9+, I recommend ([here's why](https://storybook.js.org/blog/improved-type-safety-in-storybook-7/)) typing your CSF files like this:

```suggestion
const meta = {
  title: "ui/Accordion",
  component: Accordion,
  tags: ["autodocs"],
  argTypes: {},
} satisfies Meta<typeof Accordion>;

export default meta

type Story = StoryObj<typeof meta>
```

Then each story would be typed like this:

```ts
export const Default = {
  // ...
} satisfies Story;

lloydrichards replied:
very nice, i'll try the codemod *`npx storybook migrate upgrade-deprecated-types --glob="**/*.stories.@(js|jsx|ts|tsx)"` 👍

kylegach commented on apps/www/registry/stories/badge.stories.tsx:
```diff
@@ -0,0 +1,60 @@
+/**
+ * Displays a badge or a component that looks like a badge.
+ */
+const meta: Meta<typeof Badge> = {
+  title: "ui/Badge",
+  component: Badge,
+  tags: ["autodocs"],
+  argTypes: {},
+  parameters: {
+    layout: "centered",
+  },
+}
+export default meta
+
+type Story = StoryObj<typeof Badge>
+
+/**
+ * The default form of the badge.
+ */
+export const Default: Story = {
+  render: (args) => <Badge {...args}>Badge</Badge>,
+  args: {},
+}
```
This would be more idiomatic CSF:

1. Move the "Badge" text to `args`:

   ```js
   export const Default: Story = {
     render: (args) => <Badge {...args} />,
     args: {
       children: 'Badge',
     },
   }
   ```

1. Now the `render` function simply passes all `args` along to the component, which is the same as the implicit default. Thus, it can be removed from the story definition.

1. Finally, because you're sharing the `render` across all stories, it would be most accurate to define the `children` arg value at the meta level.

Which leads to this full suggestion:

```suggestion
const meta: Meta<typeof Badge> = {
  title: "ui/Badge",
  component: Badge,
  tags: ["autodocs"],
  argTypes: {},
  args: {
    children: 'Badge',
  },
  parameters: {
    layout: "centered",
  },
}
export default meta

type Story = StoryObj<typeof Badge>

/**
 * The default form of the badge.
 */
export const Default: Story = {
  args: {},
}
```

You can then remove `render: Default.render` from the other stories.

kylegach commented on apps/www/registry/stories/button.stories.tsx:
```diff
@@ -0,0 +1,117 @@
+ * loading indicator, such as a spinner, to signify an in-progress action.
+ */
+export const Loading: Story = {
+  render: (args) => (
```
The [same guidance for Badge](https://github.com/shadcn-ui/ui/pull/1561/files#r1471776477) also applies to Button (and likely others I haven't reviewed yet; I'm not going to belabor the point).

You can still define `render` functions for stories that require a specific one (like the one I'm commenting on here) and that will override the default, implicit one.

lloydrichards replied:
amazing, i'll try implement this changes tonight 🙇‍♂️

kylegach reviewed (commented):
This is incredible work! The only comment you really need to address, imo, is the one on `.storybook/main.ts`; everything else is suggestions to bring some of the CSF definitions more in line with the idiomatic way of authoring stories, which will help folks feel like the SB docs match the generated stories here. Those suggestions are mostly about patterns, so I only commented on the first instance I saw, not every file.

If you agree with my suggestions, I'm happy to put them in place, if that's helpful.

Finally, I could not actually run the project locally (details below), so I was unable to confirm the functionality of the Storybook. But everything _looks_ good to me, afaict.

<details>
<summary>Local dev issues</summary>

I was able to clone the repo, install the deps, and run `pnpm --filter=www dev` on the `main` branch.

But when I checked out this PR's branch and tried to `pnpm install`, I got this error:

```sh
➜ pnpm install
Scope: all 4 workspace projects
Lockfile is up to date, resolution step is skipped
 WARN  Broken lockfile: no entry for '/esbuild/0.18.20' in pnpm-lock.yaml
 ERR_PNPM_ERR_PNPM_LOCKFILE_MISSING_DEPENDENCY  The lockfile is broken! Resolution step will be performed to fix it.
apps/www                                 |  WARN  deprecated rimraf@4.1.3
apps/www                                 |  WARN  Could not find preferred package /esbuild/0.18.20 in lockfile
apps/www                                 |  WARN  Could not find preferred package /esbuild/0.18.20 in lockfile
apps/www                                 |  WARN  Could not find preferred package /esbuild/0.18.20 in lockfile
apps/www                                 |  WARN  Could not find preferred package /esbuild/0.18.20 in lockfile
Packages: -14
--------------
Progress: resolved 2207, reused 2113, downloaded 0, added 0, done
node_modules/.pnpm/@swc+core@1.3.101/node_modules/@swc/core: Running postinstall scnode_modules/.pnpm/@swc+core@1.3.101/node_modules/@swc/core: Running postinstall script, done in 66ms
node_modules/.pnpm/registry.npmjs.org+esbuild@0.18.20/node_modules/esbuild: Runningnode_modules/.pnpm/registry.npmjs.org+esbuild@0.18.20/node_modules/esbuild: Running postinstall script, failed in 252ms
.../node_modules/esbuild postinstall$ node install.js
│ /Users/kylegach/projects/shadcn-ui/ui/node_modules/.pnpm/registry.npmjs.org+esb
│     throw new Error(`Expected ${JSON.stringify(versionFromPackageJSON)} but got
│           ^
│ Error: Expected "0.18.20" but got "0.17.19"
│     at validateBinaryVersion (/Users/kylegach/projects/shadcn-ui/ui/node_module
│     at /Users/kylegach/projects/shadcn-ui/ui/node_modules/.pnpm/registry.npmjs.
│ Node.js v18.18.2
└─ Failed in 252ms at /Users/kylegach/projects/shadcn-ui/ui/node_modules/.pnpm/registry.npmjs.org+esbuild@0.18.20/node_modules/esbuild
 ELIFECYCLE  Command failed with exit code 1.
```

It looks like the esbuild version isn't being found properly?

</details>

lloydrichards commented:
> If you agree with my suggestions, I'm happy to put them in place, if that's helpful.

Thanks so much! I've already gone through and made a lot of these changes.  There are a few components that rely on external props which still need the type on the variable, but i've used `satisfies` as well on these just to keep the patter going.

One thing I've notices is that with moving the render and args into the meta there are a lot of stories that are just:
```
export const Default: Story = {};
```
Is there a cleaner way to express this? Without loosing the JSDocs? Nothing too important, just something I noticed.

> Finally, I could not actually run the project locally (details below), so I was unable to confirm the functionality of the Storybook. But everything looks good to me, afaict.

And for this, there was an issue with when i did the merge conflict, but this should be solved now.  you'll need to be in the `app/www` directory to run `pnpm storybook`

kylegach commented:
You're fast!

> One thing I've notices is that with moving the render and args into the meta there are a lot of stories that are just:
> 
> ```js
> export const Default: Story = {};
> ```
> 
> Is there a cleaner way to express this? Without loosing the JSDocs? Nothing too important, just something I noticed.

If you need a story that doesn't define it's own args, parameters, etc, then that's the way. I agree it can look a bit weird, but I kinda think of it as a 1:1 relationship between the complexity of actually using the component and the verbosity of defining the story. A component's default configuration wouldn't be verbose.

> > Finally, I could not actually run the project locally (details below), so I was unable to confirm the functionality of the Storybook. But everything looks good to me, afaict.
> 
> And for this, there was an issue with when i did the merge conflict, but this should be solved now. you'll need to be in the app/www directory to run pnpm storybook

Ahh, that did it. Thanks!

lloydrichards commented:
I recently started a new project and was using the stories and noticed that many of the examples that I took from the docs when making the initial stories had a lot of cross component dependencies.  I think its important for the stories to be as simple as possible without too much opinion or dependencies so I've gone through and simplified all the stories and added removes those which didn't make sense or add anything.

Lots of the stories now follow the pattern of passing children as args and then disabling the children from the table to keep things clean.  Hopefully this still keeps the bar to entry low.

<img width="916" alt="Screenshot 2024-02-05 at 23 09 36" src="https://github.com/shadcn-ui/ui/assets/59728961/e40f8356-19ba-42cf-8ef2-2c2db36c52de">

kylegach commented on apps/www/registry/stories/accordion.stories.tsx:
```diff
@@ -15,33 +15,40 @@ const meta = {
-    </Accordion>
-  ),
+  argTypes: {
+    children: {
+      table: {
+        disable: true,
+      },
+    },
```
FYI, rather than removing `children` from the table altogether (which can imply that it's an unsupported prop), you can remove just the control like so:

```js
children: {
  control: false,
},
```

Alternatively, where appropriate (e.g. Button), you can constrain the `children` control to just a string:

```js
children: {
  control: 'text',
},
```

Docs: https://storybook.js.org/docs/api/arg-types#control (we could do a better job making this more findable 😅)

lloydrichards replied:
Which is what I ended up doing cause I regularly use the code snippet to flesh out a component and having to delete some `React.Fragment` tags will feel sloppy 😤

kylegach replied:
I so appreciate how much thought and effort you're putting into making this as useful as possible for your users!

lloydrichards commented:
I'm gonna hold my breath and say that all the stories are at a place I'm comfortable and have tested most with new projects 🤞 @shadcn I'd say that, apart from testing the CLI and making sure things are good on your end, I'm happy to wait for the next steps. 🚀

__[Storybook@chromatic](https://65cbbef5e6055247b3744fab-bqtlqhmrqt.chromatic.com/)__

I'm going to tidy up the description with some links to this Storybook that I published on Chromatic.  It would be quite easy to deploy it alongside the web app using Vercel, but will need some admin help getting the build going.  Any questions let me know and if anyone else wants to have a review or add any comments, then i'm all ears 👂

kylegach commented:
> [Storybook@chromatic](https://65cbbef5e6055247b3744fab-bqtlqhmrqt.chromatic.com/)
>
> I'm going to tidy up the description with some links to this Storybook that I published on Chromatic.

Excellent work, @lloydrichards!

Just so you know, you can [change that URL's sub-domain](https://www.chromatic.com/docs/permalinks/#build-your-own-permalink) to `https://main--65cbbef5e6055247b3744fab.chromatic.com/` and it'll always be up-to-date with the default branch's latest passing build. Those links will 404 prior to this PR being merged, of course.

ruslanguns commented:
Do you have any plan to add data table stories? https://ui.shadcn.com/docs/components/data-table

lloydrichards commented:
> Do you have any plan to add data table stories? https://ui.shadcn.com/docs/components/data-table

I think the issue I was having is that the `DataTable` is so tightly coupled with the react component you build around it, it makes it hard to show off the capabilities in an interactive way.  This was similar with the `Toast` story where I had to define a `ToastExample` component, but was at least able to pass the story args to be able to control the toast functions.

What i've been doing for my own stories is adding a `eslint-disable-next-line` and building the component within the render function.
```typescript
  render: (args) => {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const ref = useRef<ElementRef<"div">>(null);
    const { height, width, innerHeight, innerWidth, padding } =
      // eslint-disable-next-line react-hooks/rules-of-hooks
      useChartDimensions(ref, args.padding, args.padding);
    return (
      <div role="presentation" ref={ref} className="relative mb-6 size-96">
        <Chart height={height} padding={padding} width={width}>
          <g transform={`translate(${innerWidth / 2}, ${innerHeight / 2})`}>
            {args.children}
          </g>
          <DebugBoundingRect />
          <DebugSize />
        </Chart>
      </div>
    );
  },
  ```

I could do the same with the `DataTable` and `Toast`, but i'm not super comfortable about advertising this as the "correct" way of building stories 🤷‍♂️

kylegach commented:
It looks like `eslint-plugin-storybook@^0.7.0` will disable those rules of hooks for you in stories files: https://github.com/storybookjs/eslint-plugin-storybook/pull/149

lloydrichards commented:
>
>Checkbox | Name | Link to Storybook
>-- | -- | --
>... |... | ...
> ✅	|table	|[Storybook](https://65cbbef5e6055247b3744fab-bqtlqhmrqt.chromatic.com/?path=/docs/ui-table--docs)

After updating `eslint-plugin-storybook` and sitting down to make some more stories, I realised I had actually build a `Table` story with the basic elements.  @ruslanguns did you see these? or were you meaning a more feature rich story set? like with sorting?

ruslanguns commented:
Yes you are right... I totally forgot this PR was intended to improve the CLI capabilities. Thanks for your feedback.

lloydrichards commented:
No worries 😄

As a bonus (and cause I a nerd 🤓), I've gone and added stories for design tokens.  These are pulled directly from the resolved Tailwind config and so are very nice to slip into a new project (especially if you need to communicate to your designers how you're encoding their styles).

| Bonus | Name                    |                                    Link to [Storybook](https://65cbbef5e6055247b3744fab-rmvtrdhlpv.chromatic.com/) |
| :---: | ----------------------- | -----------------------------------------------------------------------------------------------------------------: |
|  ⭐️  | color - functional      |       [Storybook](https://65cbbef5e6055247b3744fab-rmvtrdhlpv.chromatic.com/?path=/story/design-color--functional) |
|  ⭐️  | color - tailwind        |         [Storybook](https://65cbbef5e6055247b3744fab-rmvtrdhlpv.chromatic.com/?path=/story/design-color--tailwind) |
|  ⭐️  | radius                  |            [Storybook](https://65cbbef5e6055247b3744fab-rmvtrdhlpv.chromatic.com/?path=/story/design-radius--core) |
|  ⭐️  | shadow                  |            [Storybook](https://65cbbef5e6055247b3744fab-rmvtrdhlpv.chromatic.com/?path=/story/design-shadow--core) |
|  ⭐️  | padding                 |           [Storybook](https://65cbbef5e6055247b3744fab-rmvtrdhlpv.chromatic.com/?path=/story/design-padding--core) |
|  ⭐️  | typography -font family | [Storybook](https://65cbbef5e6055247b3744fab-rmvtrdhlpv.chromatic.com/?path=/story/design-typography--font-family) |
|  ⭐️  | typography -font size   |   [Storybook](https://65cbbef5e6055247b3744fab-rmvtrdhlpv.chromatic.com/?path=/story/design-typography--font-size) |
|  ⭐️  | typography -font weight | [Storybook](https://65cbbef5e6055247b3744fab-rmvtrdhlpv.chromatic.com/?path=/story/design-typography--font-weight) |

So as not to complicate the CLI anymore then I need to with this, I'm going to just leave these within the `.storybook` folder for now, and once this PR is merged then I can always comeback and figure out how to add these maybe through the `npx shadcn-ui@latest init` command 🤔

For now feel free to just copy-n-paste them from the PR 🙈

kdawgwilk commented:
One minor thing that looks like its missing is support for light/dark mode theming through the `@storybook/addon-themes` addon.  Guide here https://storybook.js.org/recipes/tailwindcss#3-add-a-theme-switcher-tool

simon-v-swyftx commented on apps/www/.storybook/tokens/color.stories.tsx:
```diff
@@ -0,0 +1,125 @@
+                    ? value
+                    : hslToHex(Number(h), Number(s), Number(l))
+                  return (
+                    <div className="flex w-full flex-col pb-2">
```
missing key
```suggestion
                    <div className="flex w-full flex-col pb-2">
                    <div key={name} className="flex w-full flex-col pb-2">
```

lloydrichards commented:
> One minor thing that looks like its missing is support for light/dark mode theming through the `@storybook/addon-themes` addon. Guide here https://storybook.js.org/recipes/tailwindcss#3-add-a-theme-switcher-tool

Thanks, but for this PR i'm only looking at being able to load the stories files through the CLI with the components.  Any configuration of storybook I would leave to the user and assumed that only the basic storybook setup is available.  I've left out interaction testing (play fn) as well for the same reason 🙏

nikita-shtimenko commented:
Great PR!
Indeed, it will be great to have a Storybook support :)

ruymon commented:
@lloydrichards 

Hey! 

Just here to support your work! I've seen this PR back when it was https://github.com/shadcn-ui/ui/pull/11....

Feel free to ping me if you need anything.

lloydrichards commented:
Have just added the stories for the new Breadcrumb and InputOTP components.

@ruymon Thanks for the support! Until the vercel deploy is approved, I think the best way to support is to give a review of the work so far.  More eyes on the stories will help catch anything out of place 👍

lloydrichards commented:
> I just saw Storybook got update to version 8 recently, can we update as well?

All up to date now 🚀 Doesn't change anything for the stories since thing relied on `testing-library` or is a RSC.  But still good to keep the PR up to date and v8.0 is a welcome upgrade from Storybook since many of my own projects use RSC and the new `@storybook/test` addon

superhighfives commented:
Hey @lloydrichards—this is awesome work from everyone involved, and I'm super excited to see this land. I've been working on taking what's defined in this PR and my local shadcn/ui and storybook set up in Remix, but I'm still getting [the missing argTypes issue](https://github.com/shadcn-ui/ui/pull/1561#issuecomment-1743728518) that @kylegach responded to you on:

<img width="1066" alt="image" src="https://github.com/shadcn-ui/ui/assets/449385/64e80cb3-bd01-4e3f-9812-b5df5bbb9e38">

Wondering if there's anything special going on in the repo you're using to deploy to Chromatic, or if that's publicly available to take a look at? Figuring I must be missing some config, as I've tried the suggestion Kyle made. In the Storybook you link to at the top of this PR, the argTypes seem to render as dropdowns, etc. I am on Storybook `8`, but I tried rolling back to `7.6.17` without luck either. Maybe it's a remix thing?

Again, thanks to everyone involved in this—super cool. 😊 Happy to help with testing, however I can, too.

lloydrichards commented:
> Wondering if there's anything special going on in the repo you're using to deploy to Chromatic, or if that's publicly available to take a look at?

I don't think i'm doing anything special on the config side for Storybook, the main.ts and preview.tsx files are basically untouched.  What I had done was be explicit about the default to make sure these props show up in the control panel:
https://github.com/shadcn-ui/ui/blob/02a2dd0e3aceefdcc34bb5aef3bfe67968a3825e/apps/www/registry/stories/button.stories.tsx#L21-L25
but the argTypes i've left along for the most part.  

There are a few times where I change the display time like here in the accordion:
https://github.com/shadcn-ui/ui/blob/02a2dd0e3aceefdcc34bb5aef3bfe67968a3825e/apps/www/registry/stories/accordion.stories.tsx#L19-L23

I imagine on the chromatic side they must be creating the storybooks in a similar way, but I don't have much insight into the inner workings there 😃

superhighfives commented:
Thansk @lloydrichards—any chance you could link to the repo you're uploading to Storybook? I'd love to compare against mine to see if there's any glaring differences I might have missed.

lloydrichards commented:
> any chance you could link to the repo you're uploading to chromatic?

Its just this PR that i'm using for chromatic, I just run `npx chromatic --project-token <TOKEN>` from local development and then manually update the links 😅

simon-v-swyftx reviewed (approved):
Been using this PR's config/stories in prod for a few weeks now, still looking good!

luisalrp commented:
Hello @lloydrichards,

Thanks for building the storybook addition. I have reviewed it on your github repos and helped me to build a turbo repo starter. Would be possible to include a component that imports an Image file and loads in the storybook? Thanks!

lloydrichards commented:
> Would be possible to include a component that imports an Image file and loads in the storybook? Thanks!

Glad the PR was help to you 🥳 would you be able to elaborate on what you need here?  At the moment all the stores I've build (except the token ones) are meant to be packed with the shadcn components when adding through the CLI so if you can give the component or example it would be helpful

lloydrichards commented on apps/www/scripts/build-registry.mts:
```diff
@@ -41,9 +41,14 @@ export const Index: Record<string, any> = {
-      const resolveFiles = item.files.map(
-        (file) => `registry/${style.name}/${file}`
-      )
+      const resolveFiles = item.files.map((file) => {
+        switch (file.split(".")[1]) {
+          case "stories":
+            return `registry/${file}`
+          default:
+            return `registry/${style.name}/${file}`
+        }
+      })
```
My thinking is that a switch statement here will be more extendable in the future incase there are other supporting file types independent of the style (thinking maybe the hooks?)

lloydrichards commented on apps/www/scripts/build-registry.mts:
```diff
@@ -285,14 +290,23 @@ async function buildStyles(registry: Registry) {
-        return {
-          name: basename(file),
-          content,
+        switch (file.split(".")[1]) {
+          case "stories":
+            return {
+              name: basename(file),
+              content: readFileSync(
+                path.join(process.cwd(), "registry", file),
+                "utf8"
+              ),
+            }
+          default:
+            return {
+              name: basename(file),
+              content: readFileSync(
+                path.join(process.cwd(), "registry", style.name, file),
+                "utf8"
+              ),
+            }
```
Similar to the switch-statement in the `buildRegistry` function, there might be some room for refactoring here to make more DRY, but am not seeing a huge need with only two cases atm.

sheldonj commented:
Any help required at
This point to get this across the line?

lloydrichards commented:
I think whats needed at this point is for @shadcn to review and give any feedback on changes or additional requirements.  The PR is still part of the `next` milestone so would assume that its still planned in 🤞

Jared-Dahlke commented:
@lloydrichards I noticed in your storybook some of the components are corrupted:
![image](https://github.com/shadcn-ui/ui/assets/32936978/1fead914-c2fb-4aed-baa1-0d780ea0c76b)

I am having this same problem in my other implementation of this. it is not corrupted locally, only on the deployed instance.

lloydrichards commented:
> I noticed in your storybook some of the components are corrupted:

This would be a good question for @kylegach and the storybook team, I've ran into similar issues with using the children as an arg prop (guessing its something to do with stringifying JSX components).  

The other day I ran into a bug where if i wanted to render a ReactNode inside braces, using a template literal creates:
```typescript
const stringComponent: ReactNode = foo(); // <-- "hello"
...
return (  // <-- "([object,object])"
  <p>
    {'(${foo()}})'}
  </p>
)
```

but using string concatenation rendered correctly:
```typescript
return (  // <-- "(hello)"
  <p>
    {"("}
    {foo()}
    {")"}
  </p>
)
```

Jared-Dahlke commented:
> > I noticed in your storybook some of the components are corrupted:
> 
> This would be a good question for @kylegach and the storybook team, I've ran into similar issues with using the children as an arg prop (guessing its something to do with stringifying JSX components).
> 
> The other day I ran into a bug where if i wanted to render a ReactNode inside braces, using a template literal creates:
> 
> ```ts
> const stringComponent: ReactNode = foo(); // <-- "hello"
> ...
> return (  // <-- "([object Object])"
>   <p>
>     {'(${foo()})'}
>   </p>
> )
> ```
> 
> but using string concatenation rendered correctly:
> 
> ```ts
> return (  // <-- "(hello)"
>   <p>
>     {"("}
>     {foo()}
>     {")"}
>   </p>
> )
> ```

I figured out how to fix my implementation. I upgraded to latest v8, and also had to change all of my displayName to be raw text:
![image](https://github.com/shadcn-ui/ui/assets/32936978/48fcadcc-c084-4aac-b0bb-b7eef77bbf96)

lloydrichards commented:
Took me a little while to find time to add the stories for the New Chart components, but they're done now.  Didn't want to go too complicated so tried to use the same dataset and config for as many chart types as was appropriate.

ruslanguns commented:
> @lloydrichards I noticed in your storybook some of the components are corrupted: ![image](https://private-user-images.githubusercontent.com/32936978/336263273-1fead914-c2fb-4aed-baa1-0d780ea0c76b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIzNjQ0NDcsIm5iZiI6MTcyMjM2NDE0NywicGF0aCI6Ii8zMjkzNjk3OC8zMzYyNjMyNzMtMWZlYWQ5MTQtYzJmYi00YWVkLWJhYTEtMGQ3ODBlYTBjNzZiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzMwVDE4MjkwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM5ZjlkYjJjOTFjOTA5ZjA2Yjk3ZmM0YjFmOGFlYTM5N2FjNWNlNDQ1Nzg5ZWE1YWJmM2U3MDBlZTBhMjIzYWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.tYl_k-jXaaTwMoz2YDjo_5Zpsu6Ibr140cl5m27i1pI)
> 
> I am having this same problem in my other implementation of this. it is not corrupted locally, only on the deployed instance.

This usually happens if you are using a component that wraps your component with a memo, forwardRef, etc. and not using a displayName... Yae, I experienced the same issue in the past.

IndyVC commented:
@shadcn  any update when we can see this coming? :)

amirakbulut commented:
@shadcn Looking forward to your approvement

panzacoder commented:
> https://github.com/shadcn-ui/ui/pull/1561#issuecomment-1743728518

If anyone in 2024 is here looking for answers about pulling these components into a SB, this was the answer for me, with some revisions:

```ts

// in .storybook/main.ts config
typescript: {
    reactDocgen: 'react-docgen-typescript',
    reactDocgenTypescriptOptions: {
      shouldExtractLiteralValuesFromEnum: true,
      shouldExtractValuesFromUnion: true,
      shouldIncludePropTagMap: true,
      shouldRemoveUndefinedFromOptional: true,
      savePropValueAsString: true,

      // 👇 Adjusted prop filter
      propFilter: (prop) =>
        prop.parent
          ? !/node_modules\/(?!@radix-ui)/.test(prop.parent.fileName)
          : true
    }
  },
```

lloydrichards commented:
@shadcn I've refactored the registry changes I made previously to match the new 2.0 CLI (which has made it much easier to extend. 😘👌) mostly through extending the registry item type with `registry:story` which can be used to better select and parse the storybook files.

Have tested building the registry but not included the built files as it makes reviewing the PR quite overwhelming, but feel free to open the changes and test out the output yourself (and before committing the new output for testing I assume).

Am still motivated to get this PR merged so let me know if there is anything you need from my end or if you have some changes you would like me to tackle 🙏

0pilatos0 commented:
Would love to see this changed merged, as adding the storybooks by hand takes up some valuable time.

0pilatos0 reviewed (approved):
Great effort!

Would love to see this changed merged, as adding the storybooks by hand takes up some valuable time.

linkb15 reviewed (changes requested):
support for more deep object in colors

```js
colors: {
  primary: {
    someKey: {
      DEFAULT: "***",
      ...
      500: "***"
    }
  ...
  }
...
}

```

linkb15 commented on apps/www/.storybook/tokens/color.stories.tsx:
```diff
@@ -0,0 +1,125 @@
+      .map(([name, colors]) => {
+        return {
+          name,
+          colors: typeof colors === "string" ? { [name]: colors } : colors,
```
```suggestion
          colors: typeof colors === "string" ? { [name]: colors } : flatten(colors),
```

linkb15 commented on apps/www/.storybook/tokens/color.stories.tsx:
```diff
@@ -0,0 +1,125 @@
+      .map(([name, colors]) => {
+        return {
+          name,
+          colors: typeof colors === "string" ? { [name]: colors } : colors,
```
```suggestion
          colors: typeof colors === "string" ? { [name]: colors } : flatten(colors),
```

VoidHum commented:
@shadcn 
Any idea when this will be implemented ?

andreasss55 commented:
this should be a high priority than any other task

Saravanan90 commented:
+1 eagerly awaiting for storybook support

jeff-scifin commented on apps/www/registry/stories/separator.stories.tsx:
```diff
@@ -0,0 +1,43 @@
+/**
+ * The default form of the separator.
+ */
+export const Horizontal: Story = {
+  render: () => (
+    <div className="flex gap-2">
+      <div>Left</div>
+      <Separator orientation="vertical" className="h-auto" />
```
these two stories are backwards. this is a vertical separator and is marked as horizontal, and vice versa

lloydrichards replied:
Maybe this is a semantic thing, but when I look at [the stories](https://65cbbef5e6055247b3744fab-kvmbrnnava.chromatic.com/?path=/docs/ui-separator--docs) having them named inverse to the orientation makes more sense to me.  _Horizontally_ separated items have a _vertical_ oriented `Seperator` 🤷‍♂️

jeff-scifin replied:
The doc block says this is the default form of the separator, however `orientation="horizontal"` will be the default. 

Regardless, thanks for this PR even if it never goes in. I'm using it as a reference for adding stories to my codebase.

lloydrichards commented:
For anyone interested in a sneak-peak into the future, I've been putting together a registry for all the stories made so far.  There is a full list as well as some easy to copy commands at [registry.lloydrichards.dev](https://registry.lloydrichards.dev) but for warn that things are a wip so no judgement on styling atm 🙏 

If you want to add one of these stories to your own repo its easy, just add in the registry url for the story you want into the command below:
```bash
npx shadcn@latest add <Registry URL>
```

This will add the story and any missing components to your repo, but its important to check file and make sure the imports are correct incase you have a different ui path.

| Name                  | Registry URL                                                      |
| --------------------- | ----------------------------------------------------------------- |
| Accordion Story       | `https://registry.lloydrichards.dev/r/accordion-story.json`       |
| Alert Story           | `https://registry.lloydrichards.dev/r/alert-story.json`           |
| Alert Dialog Story    | `https://registry.lloydrichards.dev/r/alert-dialog-story.json`    |
| Aspect Ratio Story    | `https://registry.lloydrichards.dev/r/aspect-ratio-story.json`    |
| Avatar Story          | `https://registry.lloydrichards.dev/r/avatar-story.json`          |
| Badge Story           | `https://registry.lloydrichards.dev/r/badge-story.json`           |
| Breadcrumb Story      | `https://registry.lloydrichards.dev/r/breadcrumb-story.json`      |
| Button Story          | `https://registry.lloydrichards.dev/r/button-story.json`          |
| Calendar Story        | `https://registry.lloydrichards.dev/r/calendar-story.json`        |
| Card Story            | `https://registry.lloydrichards.dev/r/card-story.json`            |
| Carousel Story        | `https://registry.lloydrichards.dev/r/carousel-story.json`        |
| Chart Story           | `https://registry.lloydrichards.dev/r/chart-story.json`           |
| Checkbox Story        | `https://registry.lloydrichards.dev/r/checkbox-story.json`        |
| Collapsible Story     | `https://registry.lloydrichards.dev/r/collapsible-story.json`     |
| Command Story         | `https://registry.lloydrichards.dev/r/command-story.json`         |
| Context Menu Story    | `https://registry.lloydrichards.dev/r/context-menu-story.json`    |
| Dialog Story          | `https://registry.lloydrichards.dev/r/dialog-story.json`          |
| Drawer Story          | `https://registry.lloydrichards.dev/r/drawer-story.json`          |
| Dropdown Menu Story   | `https://registry.lloydrichards.dev/r/dropdown-menu-story.json`   |
| Form Story            | `https://registry.lloydrichards.dev/r/form-story.json`            |
| Hover Card Story      | `https://registry.lloydrichards.dev/r/hover-card-story.json`      |
| Input Story           | `https://registry.lloydrichards.dev/r/input-story.json`           |
| Input OTP Story       | `https://registry.lloydrichards.dev/r/input-otp-story.json`       |
| Label Story           | `https://registry.lloydrichards.dev/r/label-story.json`           |
| Menubar Story         | `https://registry.lloydrichards.dev/r/menubar-story.json`         |
| Navigation Menu Story | `https://registry.lloydrichards.dev/r/navigation-menu-story.json` |
| Pagination Story      | `https://registry.lloydrichards.dev/r/pagination-story.json`      |
| Popover Story         | `https://registry.lloydrichards.dev/r/popover-story.json`         |
| Progress Story        | `https://registry.lloydrichards.dev/r/progress-story.json`        |
| Radio Group Story     | `https://registry.lloydrichards.dev/r/radio-group-story.json`     |
| Resizable Story       | `https://registry.lloydrichards.dev/r/resizable-story.json`       |
| Scroll Area Story     | `https://registry.lloydrichards.dev/r/scroll-area-story.json`     |
| Select Story          | `https://registry.lloydrichards.dev/r/select-story.json`          |
| Separator Story       | `https://registry.lloydrichards.dev/r/separator-story.json`       |
| Sheet Story           | `https://registry.lloydrichards.dev/r/sheet-story.json`           |
| Skeleton Story        | `https://registry.lloydrichards.dev/r/skeleton-story.json`        |
| Slider Story          | `https://registry.lloydrichards.dev/r/slider-story.json`          |
| Sonner Story          | `https://registry.lloydrichards.dev/r/sonner-story.json`          |
| Switch Story          | `https://registry.lloydrichards.dev/r/switch-story.json`          |
| Table Story           | `https://registry.lloydrichards.dev/r/table-story.json`           |
| Tabs Story            | `https://registry.lloydrichards.dev/r/tabs-story.json`            |
| Textarea Story        | `https://registry.lloydrichards.dev/r/textarea-story.json`        |
| Toast Story           | `https://registry.lloydrichards.dev/r/toast-story.json`           |
| Toggle Story          | `https://registry.lloydrichards.dev/r/toggle-story.json`          |
| Toggle Group Story    | `https://registry.lloydrichards.dev/r/toggle-group-story.json`    |
| Tooltip Story         | `https://registry.lloydrichards.dev/r/tooltip-story.json`         |

I've made the [repo](https://github.com/lloydrichards/shadcn-storybook-registry) pubic so if you have any suggestions or issues, feel free to add them there 👍

lloydrichards commented:
Firstly I want to thank everyone who has subscribes and been part of the conversation to try and get this PR merged in 🫶  Its been a lot of fun the last 18 months keeping up with the merge commits and tweaking the stories.

That said, and with the release of [v2.3.0](https://github.com/shadcn-ui/ui/releases/tag/shadcn%402.3.0), it more evident then every that this PR is not going to be merged and instead the focus is to build a robust ecosystem of components rather then a monolithic repository.  I personally feel this is the correct path forward for shadcn/ui as it elevates the pressure of constant component and block requests and shifts responsibility of maintenance towards registries.

## Next steps for Stories

As mentioned in the previous comment 👆 I've used the [new registry documentations](https://ui.shadcn.com/docs/registry) to build out a place for the stories I've build and will continue to maintain and use them from here.  If you haven't already, check out the website for lists of cli commands or visit the [repo](https://github.com/lloydrichards/shadcn-storybook-registry) if you want to help contribute through issues or PRs

**https://registry.lloydrichards.dev/**

This might also be a good time for Storybook (@kylegach 👀 ) to fork or take over this initiative with their own registry, but no pressure or hard feelings either way.  Just happy to have been part of the journey 🫡 

## Next step for this PR

This is where it would be great to get the final input from @shadcn (but I know is super busy with Tailwind v4 issues atm) but my thoughts are that this PR is now stale and can be closed to help reduce the entropy a bit in the open PRs.  Unless i hear otherwise, I'll close the PR myself in ~1 week 🙏 

Thanks again for the support and look forward to helping out in the repo in other ways in the future ❤️

adamfratino commented:
@lloydrichards amazing contribution to the community, regardless of merge status. thank you for all your good work! 👏👏👏

superhighfives commented:
Seconding that—I've been following this issue, I think since its inception, and it's a testament to your commitment to open source and to the community that you've stuck with it, @lloydrichards. Awesome work, much appreciated. 🙌

kylegach commented:
Thank you, Lloyd. Your work creating and maintaining this PR has been incredible. I especially appreciate how receptive you were to making sure anyone using this would get the latest Storybook best practices.

I've raised this with the Storybook team, and I'll let you know what comes of that as soon as we're able.

kylegach commented:
@lloydrichards — Can you please email me at kyle@chromatic.com, so we can discuss next steps?

(Sorry for the notification spam, everyone else. 🙃)
